### ISSUE TITLE
Text Search for query 'age' returns every tool and user message

### ISSUE THREAD

**Issue by madgrizzle** [cpacker/MemGPT#1282]
**Describe the bug**
It appears that the text_search function here:
https://github.com/cpacker/MemGPT/blob/eaed123af8584d139879e56a23e5bc6237bdfcde/memgpt/agent_store/db.py#L404
performs a query on the "text" field stored in the SQL database.  

From what I can tell, the "text" field is a json in text format that contains the key "message".  When I ask MemGPT how old I am, it performs a query based on the keyword "age" and hits on "message" and therefore returns every result.

![image](https://github.com/cpacker/MemGPT/assets/5055291/9221abd1-a5f1-4e1a-85ca-5603708db681)

**Additional context**
I don't readily know how, but the text_search function needs to search meaningful portion of the message, but I think the issue is bigger than just this particular case.

**Comment by yoadsn** [comment]
The current implementation of `query_text` in `SQLStorageConnector` is generic and works for both Postgres and SQLite db dialects. It performs a very basic text matching as the code acknowledges.

```
# todo: make fuzz https://stackoverflow.com/questions/42388956/create-a-full-text-search-index-with-sqlalchemy-on-postgresql/42390204#42390204
```

The solution to that problem could look like this:
- Create a specialized `query_text` implementation for in `PostgresStorageConnector`
- Make use of the full text search abilities Postgres provides
- In SQLite, perhaps find a way to use semantic vector rerank on top of retrieved results (using Chroma, and the embedding model at-hand) but admittedly, the first result-set is a poor selection in the first place.

Ideally, textual search would use a dedicated (local) textual search engine which is tailored for that purpose. Such engine could be useful for many other flows that could help conversation injection with reduced token/embedding costs.

Would you be able to work on a PR for this?

**Comment by madgrizzle** [comment]
From what I read, FTS is a good way to go.  I don't know that I'm a strong enough programmer to make it work but will play around with it.  But I ran into two other issues to overcome during testing.. 

1) When I ask it "how old am I?" sometimes it queries with "age" but sometimes also "Chad's age".  A better query would be ("Chad" AND "age") or maybe ("Chad" NEAR "age").  Landing a hit on "Chad's age" seem's challenging unless it is something that the LLM actually returns as an inner thought, but inner thoughts aren't captured in the conversation history.  Any reason they cannot stored as well for querying?  It seems this might be rather desirable as an LLM might "normalize" an input of information in the same way.. If I say "Chad is 12 year's old" it my have an inner thought of "Chad's age is 12" and then when i ask "How old is Chad" it may then query on "Chad's age" and make a hit on that inner thought.

2) The text_search includes system message and if there's a hit inside it that for some reason the LLM didn't pick up on during the initial processing of the request, then the text_search query results will include the system message.  This will almost always create an issue where the prompt exceeds the context size because the system messages are rather large.  I added a filter to exclude system messages and it seems work better.  Is there any reason system message should be searched during conversation_search?
